test/syncenginetestutils.h

```diff
@@ -0,0 +1,658 @@
+        QFile file{_rootDir.filePath(relativePath)};
+        QVERIFY(file.exists());
+        file.open(QFile::ReadWrite);
+        QByteArray contents = file.read(1);
+        file.seek(file.size());
+        file.write(contents);
+    }
+    void mkdir(const QString &relativePath) override {
+        _rootDir.mkpath(relativePath);
+    }
+};
+
+class FileInfo : public FileModifier
+{
+public:
+    static FileInfo A12_B12_C12_S12() {
+        FileInfo fi{QString{}, {
+            {QStringLiteral("A"), {
+                {QStringLiteral("a1"), 4},
+                {QStringLiteral("a2"), 4}
+            }},
+            {QStringLiteral("B"), {
+                {QStringLiteral("b1"), 16},
+                {QStringLiteral("b2"), 16}
+            }},
+            {QStringLiteral("C"), {
+                {QStringLiteral("c1"), 24},
+                {QStringLiteral("c2"), 24}
+            }},
+        }};
+        FileInfo sharedFolder{QStringLiteral("S"), {
+            {QStringLiteral("s1"), 32},
+            {QStringLiteral("s2"), 32}
+        }};
+        sharedFolder.isShared = true;
+        sharedFolder.children[QStringLiteral("s1")].isShared = true;
+        sharedFolder.children[QStringLiteral("s2")].isShared = true;
+        fi.children.insert(sharedFolder.name, std::move(sharedFolder));
+        return fi;
+    }
+
+    FileInfo() = default;
+    FileInfo(const QString &name) : name{name} { }
+    FileInfo(const QString &name, qint64 size) : name{name}, isDir{false}, size{size} { }
+    FileInfo(const QString &name, qint64 size, char contentChar) : name{name}, isDir{false}, size{size}, contentChar{contentChar} { }
+    FileInfo(const QString &name, const std::initializer_list<FileInfo> &children) : name{name} {
+        QString p = path();
+        for (const auto &source : children) {
+            auto &dest = this->children[source.name] = source;
+            dest.parentPath = p;
+        }
+    }
+
+    void remove(const QString &relativePath) override {
+        const QStringList pathComponents = relativePath.split('/', QString::SkipEmptyParts);
+        FileInfo *parent = findInvalidatingEtags(pathComponents.mid(0, pathComponents.size() - 1));
+        Q_ASSERT(parent);
+        parent->children.erase(std::find_if(parent->children.begin(), parent->children.end(),
+                [&pathComponents](const FileInfo &fi){ return fi.name == pathComponents.last(); }));
+    }
+    void insert(const QString &relativePath, qint64 size = 64, char contentChar = 'W') override {
+        create(relativePath, size, contentChar);
+    }
+    void setContents(const QString &relativePath, char contentChar) override {
+        FileInfo *file = findInvalidatingEtags(relativePath.split('/', QString::SkipEmptyParts));
+        Q_ASSERT(file);
+        file->contentChar = contentChar;
+    }
+    void appendByte(const QString &relativePath) override {
+        FileInfo *file = findInvalidatingEtags(relativePath.split('/', QString::SkipEmptyParts));
```

Comment:
Minor: `.split('/', QString::SkipEmptyParts)` and `.mid(0, pathComponents.size() - 1)` happen often enough that they might be functions.